### PR TITLE
Access control against Keycloak (KeycloakLoginModule)

### PR DESCRIPTION
**Pull request by nebilbenmabrouk**
- Add a **JAAS** **Keycloak** **Login Module**
- This module is based on the code of the JAAS Adapter of Keycloak called [DirectAccessGrantsLoginModule](https://www.keycloak.org/docs/latest/securing_apps/#_jaas_adapter)
- It authenticates users against Keycloak and acquires their information (including their **roles**)
- It uses the **OpenID Connect (OIDC)** protocol to communicate with [Keycloak REST endpoints for OIDC](https://www.keycloak.org/docs/latest/securing_apps/#endpoints) 
- It uses the configuration file located under **config/keycloak.cfg**
- It uses the latest version of the **keycloak-adapter-core dependency** (_version 24.0.2_) and supports previous releases of Keycloak server.

### CODE REVIEW

File: common/common-client/build.gradle [hunk 1]
```diff
@@ -5,6 +5,8 @@ dependencies {
     compile "org.objectweb.proactive:programming-core:${programmingVersion}"
     compile "org.objectweb.proactive:programming-extension-annotation:${programmingVersion}"
 
+    compile "org.keycloak:keycloak-adapter-core:24.0.2"
```
Comment (fviale) [inline common/common-client/build.gradle:8]: can we please check together that this dependency does not add conflicts with our existing libraries ?

File: common/common-client/src/main/java/org/ow2/proactive/authentication/keycloak/KeycloakOidcProperties.java [hunk 2]
```diff
@@ -0,0 +1,201 @@
...
+     * true or false */
+    public static final String FALLBACK_USER_AUTH = "pa.keycloak.authentication.fallback";
+
+    /** group fall back property, check user group membership group file if user is not found in corresponding Keycloak group.
+     * true or false */
+    public static final String FALLBACK_GROUP_MEMBERSHIP = "pa.keycloak.group.membership.fallback";
+
+    /** tenant fall back property, check user tenant membership tenant file if user is not found in LDAP or if tenant Keycloak attribute is not defined.
+     * true or false */
+    public static final String FALLBACK_TENANT_MEMBERSHIP = "pa.keycloak.tenant.membership.fallback";
```
Comment (fviale) [inline common/common-client/src/main/java/org/ow2/proactive/authentication/keycloak/KeycloakOidcProperties.java:94]: As there is no property that defines a tenant attribute in the keycloack configuration, I don't think this is useful.

We might consider though what would be multi-tenant when using keycloak

File: common/common-client/build.gradle (hunk 1 above)
Comment (nebilbenmabrouk) [reply common/common-client/build.gradle:8]: I already checked and all Keycloak transitive dependencies are green.

- This is the list of dependencies induced by **keycloak-adapter-core:24.0.2**:

```
|    |    +--- org.keycloak:keycloak-adapter-core:24.0.2
|    |    |    \--- org.keycloak:keycloak-crypto-default:24.0.2
|    |    |         +--- org.keycloak:keycloak-core:24.0.2
|    |    |         |    +--- org.keycloak:keycloak-common:24.0.2
|    |    |         |    |    \--- jakarta.activation:jakarta.activation-api:2.1.2
|    |    |         |    +--- com.fasterxml.jackson.core:jackson-core:2.16.1
|    |    |         |    +--- com.fasterxml.jackson.core:jackson-databind:2.16.1
|    |    |         |    |    +--- com.fasterxml.jackson.core:jackson-annotations:2.16.1
|    |    |         |    |    \--- com.fasterxml.jackson.core:jackson-core:2.16.1
|    |    |         |    \--- org.eclipse.microprofile.openapi:microprofile-openapi-api:3.1.1
|    |    |         +--- org.keycloak:keycloak-server-spi:24.0.2
|    |    |         +--- org.keycloak:keycloak-server-spi-private:24.0.2
|    |    |         |    \--- org.apache.httpcomponents:httpclient:4.5.14 (*)
|    |    |         +--- org.bouncycastle:bcprov-jdk18on:1.77
|    |    |         \--- org.bouncycastle:bcpkix-jdk18on:1.77
|    |    |              +--- org.bouncycastle:bcprov-jdk18on:1.77
|    |    |              \--- org.bouncycastle:bcutil-jdk18on:1.77
|    |    |                   \--- org.bouncycastle:bcprov-jdk18on:1.77

```

- This is the list of current vulnerabilites in the **Scheduling project**. None of them is induced by Keycloak :

image

File: common/common-client/src/main/java/org/ow2/proactive/authentication/keycloak/KeycloakOidcProperties.java (hunk 2 above)
Comment (nebilbenmabrouk) [reply common/common-client/src/main/java/org/ow2/proactive/authentication/keycloak/KeycloakOidcProperties.java:94]: So should I remove it or keep it for a later use (_when we consider multi-tenant based on keycloak_) ?

**Review by fviale — APPROVED** [review]

**Comment by nebilbenmabrouk** [comment]
re-test this please

**Pull request merged**